File: app/src/main/java/com/bongorian/signa1/EffectParameters.java
package com.bongorian.signa1;

import java.util.*;

/** Immutable named UI macros. There is no strength/param1/param2/seed stride or physical ABI. */
final class EffectParameters {
    private final Map<Integer,Map<String,Float>> values;
    private final Map<Integer,Long> identities,eventIdentities;
    private final Map<Integer,Map<String,Float>> overrides;
    private EffectParameters(Map<Integer,Map<String,Float>> values,Map<Integer,Long> identities){this(values,identities,Collections.emptyMap());}
    private EffectParameters(Map<Integer,Map<String,Float>> values,Map<Integer,Long> identities,Map<Integer,Map<String,Float>> overrides){this(values,identities,overrides,Collections.emptyMap());}
    private EffectParameters(Map<Integer,Map<String,Float>> values,Map<Integer,Long> identities,Map<Integer,Map<String,Float>> overrides,Map<Integer,Long> eventIdentities){
        this.eventIdentities=Collections.unmodifiableMap(new LinkedHashMap<>(eventIdentities));
        Map<Integer,Map<String,Float>> copy=new LinkedHashMap<>();
        values.forEach((id,p)->copy.put(id,Collections.unmodifiableMap(new LinkedHashMap<>(p))));
        Map<Integer,Map<String,Float>> manual=new LinkedHashMap<>();overrides.forEach((id,p)->manual.put(id,Collections.unmodifiableMap(new LinkedHashMap<>(p))));this.overrides=Collections.unmodifiableMap(manual);
        this.values=Collections.unmodifiableMap(copy);this.identities=Collections.unmodifiableMap(new LinkedHashMap<>(identities));
    }
    static float unit(float value){return Float.isFinite(value)?Math.max(0,Math.min(1,value)):0;}
    static EffectParameters defaults(){
        Map<Integer,Map<String,Float>> values=new LinkedHashMap<>();Map<Integer,Long> identities=new LinkedHashMap<>();
        for(int id:Effects.ORDER){Map<String,Float> controls=new LinkedHashMap<>();for(Effects.Control c:Effects.CONTROLS[id])controls.put(c.key,c.initial);values.put(id,controls);identities.put(id,0x51a1L+id*1000003L);}
        return new EffectParameters(values,identities);
    }
    float get(int id,String key){Float v=values.get(id).get(key);if(v==null)throw new IllegalArgumentException("Unknown control: "+id+" / "+key);return v;}
    long identity(int id){return identities.get(id);}
    EffectParameters with(int id,String key,float value){
        get(id,key);Map<Integer,Map<String,Float>> copy=new LinkedHashMap<>(values);Map<String,Float> controls=new LinkedHashMap<>(values.get(id));controls.put(key,unit(value));copy.put(id,controls);return new EffectParameters(copy,identities,overrides,eventIdentities);
    }
    EffectParameters reseed(int id,long seed){Map<Integer,Long> copy=new LinkedHashMap<>(identities);if(!copy.containsKey(id))throw new IllegalArgumentException("Fault ID");copy.put(id,seed);return new EffectParameters(values,copy,overrides,eventIdentities);}
    EffectParameters reset(int id){EffectParameters result=clearOverrides(id);for(Effects.Control c:Effects.CONTROLS[id])result=result.with(id,c.key,c.initial);return result;}
    boolean fixedEventIdentity(int id){return eventIdentities.containsKey(id);}
    long eventIdentity(int id,long fallback){return eventIdentities.getOrDefault(id,fallback);}
    EffectParameters eventIdentity(int id,Long value){if(!FaultParameters.incidents(id))throw new IllegalArgumentException("No event generator");Map<Integer,Long> copy=new LinkedHashMap<>(eventIdentities);if(value==null)copy.remove(id);else copy.put(id,value);return new EffectParameters(values,identities,overrides,copy);}
    Map<String,Float> overrides(int id){return overrides.getOrDefault(id,Collections.emptyMap());}
    boolean manual(int id,String key){return overrides(id).containsKey(key);}
    float resolved(int id,String key,float automatic){return overrides(id).getOrDefault(key,automatic);}
    EffectParameters override(int id,String key,float value){FaultParameters.spec(id,key).validate(value);Map<Integer,Map<String,Float>> copy=new LinkedHashMap<>(overrides);Map<String,Float> controls=new LinkedHashMap<>(overrides(id));controls.put(key,value);copy.put(id,controls);return new EffectParameters(values,identities,copy,eventIdentities);}
    EffectParameters automatic(int id,String key){FaultParameters.spec(id,key);Map<Integer,Map<String,Float>> copy=new LinkedHashMap<>(overrides);Map<String,Float> controls=new LinkedHashMap<>(overrides(id));controls.remove(key);if(controls.isEmpty())copy.remove(id);else copy.put(id,controls);return new EffectParameters(values,identities,copy,eventIdentities);}
    EffectParameters clearOverrides(int id){Map<Integer,Map<String,Float>> copy=new LinkedHashMap<>(overrides);copy.remove(id);Map<Integer,Long> events=new LinkedHashMap<>(eventIdentities);events.remove(id);return new EffectParameters(values,identities,copy,events);}
    String encode(){StringBuilder s=new StringBuilder();for(int id:Effects.ORDER){if(id==0)continue;s.append(';').append(id).append(':').append(identity(id));for(Effects.Control c:Effects.CONTROLS[id])s.append(',').append(c.key).append('=').append(get(id,c.key));if(fixedEventIdentity(id))s.append(",@eventIdentity=").append(eventIdentity(id,0));for(Map.Entry<String,Float> entry:new TreeMap<>(overrides(id)).entrySet())s.append(",@").append(entry.getKey()).append('=').append(entry.getValue());}return s.toString();}
    static EffectParameters decode(String text){
        EffectParameters p=defaults();Set<Integer> seen=new HashSet<>();
        for(String group:text.split(";")){if(group.isEmpty())continue;String[] entries=group.split(",");String[] head=entries[0].split(":");if(head.length!=2)throw new IllegalArgumentException("Fault identity");int id=Integer.parseInt(head[0]);
            if(id<=0||id>=Effects.NAMES.length||!seen.add(id))throw new IllegalArgumentException("Fault ID");p=p.reseed(id,Long.parseLong(head[1]));Set<String> keys=new HashSet<>();
            for(int i=1;i<entries.length;i++){String[] pair=entries[i].split("=");if(pair.length!=2||!keys.add(pair[0]))throw new IllegalArgumentException("Fault control");if(pair[0].equals("@eventIdentity")){p=p.eventIdentity(id,Long.valueOf(pair[1]));continue;}float value=Float.parseFloat(pair[1]);if(pair[0].startsWith("@")){p=p.override(id,pair[0].substring(1),value);}else{if(!Float.isFinite(value)||value<0||value>1)throw new IllegalArgumentException("Control bounds");p=p.with(id,pair[0],value);}}
            if(keys.stream().filter(key->!key.startsWith("@")).count()!=Effects.CONTROLS[id].length)throw new IllegalArgumentException("Missing controls");
        }
        if(seen.size()!=Effects.NAMES.length-1)throw new IllegalArgumentException("Missing faults");return p;
    }
    String describe(int[] ids){StringBuilder s=new StringBuilder();for(int id:ids){s.append(" | ").append(Effects.name(id)).append(" identity=").append(identity(id));for(Effects.Control c:Effects.CONTROLS[id])s.append(' ').append(c.key).append('=').append(get(id,c.key));if(fixedEventIdentity(id))s.append(" eventIdentity=").append(eventIdentity(id,0));if(!overrides(id).isEmpty())s.append(" overrides=").append(overrides(id));}return s.toString();}
}

File: app/src/main/kotlin/com/bongorian/signa1/EffectParameters.kt
package com.bongorian.signa1

import java.util.TreeMap
import kotlin.math.max
import kotlin.math.min

/** Immutable named UI macros. There is no strength/param1/param2/seed stride or physical ABI. */
internal class EffectParameters
private constructor(
    values: Map<Int, Map<String, Float>>,
    identities: Map<Int, Long>,
    overrides: Map<Int, Map<String, Float>> = emptyMap<Int, Map<String, Float>>(),
    eventIdentities: Map<Int, Long> = emptyMap<Int, Long>(),
) {
    private val values =
        values.mapValues { (_, controls) -> controls.immutableCopy() }.immutableCopy()
    private val identities = identities.immutableCopy()
    private val eventIdentities = eventIdentities.immutableCopy()
    private val overrides =
        overrides.mapValues { (_, controls) -> controls.immutableCopy() }.immutableCopy()

    fun get(id: Int, key: String): Float {
        val v = values.getValue(id)[key]
        requireNotNull(v) { "Unknown control: " + id + " / " + key }
        return v
    }

    fun identity(id: Int): Long {
        return identities.getValue(id)
    }

    fun with(id: Int, key: String, value: Float): EffectParameters {
        get(id, key)
        val copy: MutableMap<Int, Map<String, Float>> =
            LinkedHashMap<Int, Map<String, Float>>(values)
        val controls: MutableMap<String, Float> = LinkedHashMap<String, Float>(values.get(id))
        controls.put(key, unit(value))
        copy.put(id, controls)
        return EffectParameters(copy, identities, overrides, eventIdentities)
    }

    fun reseed(id: Int, seed: Long): EffectParameters {
        val copy: MutableMap<Int, Long> = LinkedHashMap<Int, Long>(identities)
        require(copy.containsKey(id)) { "Fault ID" }
        copy.put(id, seed)
        return EffectParameters(values, copy, overrides, eventIdentities)
    }

    fun reset(id: Int): EffectParameters {
        var result = clearOverrides(id)
        for (c in Effects.CONTROLS[id]) result = result.with(id, c.key, c.initial)
        return result
    }

    fun fixedEventIdentity(id: Int): Boolean {
        return eventIdentities.containsKey(id)
    }

    fun eventIdentity(id: Int, fallback: Long): Long {
        return eventIdentities.getOrDefault(id, fallback)
    }

    fun withEventIdentity(id: Int, value: Long?): EffectParameters {
        require(FaultParameters.incidents(id)) { "No event generator" }
        val copy: MutableMap<Int, Long> = LinkedHashMap<Int, Long>(eventIdentities)
        if (value == null) copy.remove(id) else copy.put(id, value)
        return EffectParameters(values, identities, overrides, copy)
    }

    fun overrides(id: Int): Map<String, Float> {
        return overrides.getOrDefault(id, mutableMapOf<String, Float>())
    }

    fun manual(id: Int, key: String): Boolean {
        return overrides(id).containsKey(key)
    }

    fun resolved(id: Int, key: String, automatic: Float): Float {
        return overrides(id).getOrDefault(key, automatic)
    }

    fun override(id: Int, key: String, value: Float): EffectParameters {
        FaultParameters.spec(id, key).validate(value)
        val copy: MutableMap<Int, Map<String, Float>> =
            LinkedHashMap<Int, Map<String, Float>>(overrides)
        val controls: MutableMap<String, Float> = LinkedHashMap<String, Float>(overrides(id))
        controls.put(key, value)
        copy.put(id, controls)
        return EffectParameters(values, identities, copy, eventIdentities)
    }

    fun automatic(id: Int, key: String): EffectParameters {
        FaultParameters.spec(id, key)
        val copy: MutableMap<Int, Map<String, Float>> =
            LinkedHashMap<Int, Map<String, Float>>(overrides)
        val controls: MutableMap<String, Float> = LinkedHashMap<String, Float>(overrides(id))
        controls.remove(key)
        if (controls.isEmpty()) copy.remove(id) else copy.put(id, controls)
        return EffectParameters(values, identities, copy, eventIdentities)
    }

    fun clearOverrides(id: Int): EffectParameters {
        val copy: MutableMap<Int, Map<String, Float>> =
            LinkedHashMap<Int, Map<String, Float>>(overrides)
        copy.remove(id)
        val events: MutableMap<Int, Long> = LinkedHashMap<Int, Long>(eventIdentities)
        events.remove(id)
        return EffectParameters(values, identities, copy, events)
    }

    fun encode(): String {
        val s = StringBuilder()
        for (id in Effects.ORDER) {
            if (id == 0) continue
            s.append(';').append(id).append(':').append(identity(id))
            for (c in Effects.CONTROLS[id]) s.append(',')
                .append(c.key)
                .append('=')
                .append(get(id, c.key))
            if (fixedEventIdentity(id)) s.append(",@eventIdentity=").append(eventIdentity(id, 0))
            for (entry in TreeMap<String, Float>(overrides(id)).entries) s.append(",@")
                .append(entry.key)
                .append('=')
                .append(entry.value)
        }
        return s.toString()
    }

    fun describe(ids: IntArray): String {
        val s = StringBuilder()
        for (id in ids) {
            s.append(" | ").append(Effects.name(id)).append(" identity=").append(identity(id))
            for (c in Effects.CONTROLS[id]) s.append(' ')
                .append(c.key)
                .append('=')
                .append(get(id, c.key))
            if (fixedEventIdentity(id)) s.append(" eventIdentity=").append(eventIdentity(id, 0))
            if (!overrides(id).isEmpty()) s.append(" overrides=").append(overrides(id))
        }
        return s.toString()
    }

    companion object {
        fun unit(value: Float): Float {
            return if (value.isFinite()) max(0f, min(1f, value)) else 0f
        }

        fun defaults(): EffectParameters {
            val values: MutableMap<Int, MutableMap<String, Float>> =
                LinkedHashMap<Int, MutableMap<String, Float>>()
            val identities: MutableMap<Int, Long> = LinkedHashMap<Int, Long>()
            for (id in Effects.ORDER) {
                val controls: MutableMap<String, Float> = LinkedHashMap<String, Float>()
                for (c in Effects.CONTROLS[id]) controls.put(c.key, c.initial)
                values.put(id, controls)
                identities.put(id, 0x51a1L + id * 1000003L)
            }
            return EffectParameters(values, identities)
        }

        fun decode(text: String): EffectParameters {
            var p: EffectParameters = defaults()
            val seen: MutableSet<Int> = HashSet<Int>()
            for (group in text.split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()) {
                if (group.isEmpty()) continue
                val entries =
                    group.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val head =
                    entries[0].split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                require(head.size == 2) { "Fault identity" }
                val id = head[0].toInt()
                require(!(id <= 0 || id >= Effects.NAMES.size || !seen.add(id))) { "Fault ID" }
                p = p.reseed(id, head[1].toLong())
                val keys: MutableSet<String> = HashSet<String>()
                for (i in 1..<entries.size) {
                    val pair =
                        entries[i]
                            .split("=".toRegex())
                            .dropLastWhile { it.isEmpty() }
                            .toTypedArray()
                    require(!(pair.size != 2 || !keys.add(pair[0]))) { "Fault control" }
                    if (pair[0] == "@eventIdentity") {
                        p = p.withEventIdentity(id, pair[1].toLong())
                        continue
                    }
                    val value = pair[1].toFloat()
                    if (pair[0].startsWith("@")) {
                        p = p.override(id, pair[0].substring(1), value)
                    } else {
                        require(!(!value.isFinite() || value < 0 || value > 1)) { "Control bounds" }
                        p = p.with(id, pair[0], value)
                    }
                }
                require(keys.count { !it.startsWith("@") } == Effects.CONTROLS[id].size) {
                    "Missing controls"
                }
            }
            require((1..Effects.CRT).all { it in seen }) { "Missing legacy faults" }
            return p
        }
    }
}
